colab - https://colab.research.google.com/drive/1qBBkhh_p3F67xK2ezcurmjDuEsFwh48g?authuser=0#scrollTo=-Dwa_X-VVkft

In [1]:
from google.colab import userdata
OPENAI_API_KEY = userdata.get('OPENAI_API_KEY')

In [2]:
!pip install -q dspy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 291.3/291.3 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.7/119.7 kB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 2.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 95.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 404.7/404.7 kB 26.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.1/278.1 kB 10.1 MB/s eta 0:00:00


In [3]:
import dspy

lm = dspy.LM(
    model="openai/gpt-4.1-mini",
    temperature=1,
    api_key=OPENAI_API_KEY
)
dspy.configure(lm=lm)

In [4]:
lm("hi")

/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content='Hello! H...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(


['Hello! How can I assist you today?']

In [7]:
import dspy
from datasets import load_dataset

def dataset():
  train = load_dataset("AI-MO/aimo-validation-aime")["train"]
  train = [
      dspy.Example({
          "problem": x["problem"],
          "solution": x["solution"],
          "answer": x["answer"]
      }).with_inputs("problem")
      for x in train
  ]

  import random
  random.Random(0).shuffle(train)
  train_len = len(train)

  test = load_dataset("MathArena/aime_2025")["train"]
  test = [
      dspy.Example({
          "problem": x["problem"],
          "answer": x["answer"]
      }).with_inputs("problem")
      for x in test
  ]

  train_set = train[:int(0.5*train_len)]
  val_set = train[int(0.5*train_len):]
  test_set = test*5

  return train_set, val_set, test_set



# quick notes
- we are using 2 different datasets for train and test (both related to math)
- we use a validation set from train to imrprove optimisation
- .with_inputs("problem") tells dspy to only send problem to llm and hide the solution / answer
- train and val sets are half of the og train dataset
- test set is duplicated to check llm performance across 5 different responses of same query

In [8]:
train_set, val_set, test_set = dataset()
len(train_set), len(val_set), len(test_set)

(45, 45, 150)

In [9]:
print("Problem:\n")
print(train_set[0]['problem'])
print("\n\nSolution:\n")
print(train_set[0]['solution'])
print("\n\nAnswer:\n")
print(train_set[0]['answer'])

Problem:

In isosceles trapezoid $ABCD$, parallel bases $\overline{AB}$ and $\overline{CD}$ have lengths $500$ and $650$, respectively, and $AD=BC=333$. The angle bisectors of $\angle{A}$ and $\angle{D}$ meet at $P$, and the angle bisectors of $\angle{B}$ and $\angle{C}$ meet at $Q$. Find $PQ$.


Solution:

We have the following diagram:

Let $X$ and $W$ be the points where $AP$ and $BQ$ extend to meet $CD$, and $YZ$ be the height of $\triangle AZB$. As proven in Solution 2, triangles $APD$ and $DPW$ are congruent right triangles. Therefore, $AD = DW = 333$. We can apply this logic to triangles $BCQ$ and $XCQ$ as well, giving us $BC = CX = 333$. Since $CD = 650$, $XW = DW + CX - CD = 16$.
Additionally, we can see that $\triangle XZW$ is similar to $\triangle PQZ$ and $\triangle AZB$. We know that $\frac{XW}{AB} = \frac{16}{500}$. So, we can say that the height of the triangle $AZB$ is $500u$ while the height of the triangle $XZW$ is $16u$. After that, we can figure out the distance fro

In [16]:
class GenerateResponse(dspy.Signature):
  """Solve the problem and provide the answer in the correct format."""
  problem = dspy.InputField()
  answer = dspy.OutputField()

program = dspy.ChainOfThought(GenerateResponse)

In [11]:
def metric(example, prediction, trace=None, pred_name=None, pred_trace=None):
  correct_answer=int(example["answer"])
  try:
    llm_answer = int(prediction.answer)
  except ValueError as e:
    return 0
  return int(correct_answer == llm_answer)

In [12]:
import dspy

evaluate = dspy.Evaluate(
    devset=test_set,
    metric=metric,
    num_threads=32,
    display_table=True,
    display_progress=True
)

evaluate(program)

Average Metric: 0.00 / 2 (0.0%):   1%|          | 1/150 [00:17<43:34, 17.54s/it]

/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content='[[ ## re...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(


Average Metric: 1.00 / 3 (33.3%):   2%|▏         | 3/150 [00:23<16:26,  6.71s/it]

/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content="[[ ## re...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(


Average Metric: 67.00 / 150 (44.7%): 100%|██████████| 150/150 [05:18<00:00,  2.12s/it]

2026/01/16 20:54:50 INFO dspy.evaluate.evaluate: Average Metric: 67 / 150 (44.7%)


problem  \
0    Find the sum of all integer bases $b>9$ for which $17_b$ is a divi...   
1    On $\triangle ABC$ points $A, D, E$, and $B$ lie in that order on ...   
2    The 9 members of a baseball team went to an ice-cream parlor after...   
3    Find the number of ordered pairs $(x,y)$, where both $x$ and $y$ a...   
4    There are $8!= 40320$ eight-digit positive integers that use each ...   
..                                                                     ...   
145  Let $S$ be the set of vertices of a regular $24$-gon. Find the num...   
146  Let $A_1 A_2 A_3 \ldots A_{11}$ be an $11$-sided non-convex simple...   
147  Let $x_1, x_2, x_3, \ldots$ be a sequence of rational numbers defi...   
148  Let $\triangle ABC$ be a right triangle with $\angle A = 90^\circ$...   
149  There are exactly three positive real numbers $k$ such that the fu...   

     example_answer  \
0                70   
1               588   
2                16   
3               117   
4               279   
..              ...   
145             113   
146              19   
147             248   
148             104   
149             240   

                                                                 reasoning  \
0    First, let's understand the problem in detail: We are given two nu...   
1    First, let's analyze the given configuration and the notation care...   
2    We have 9 players, each chooses one flavor: chocolate (C), vanilla...   
3    We need to find the number of integer pairs \((x,y)\) with \(-100 ...   
4    We are asked to consider all 8-digit positive integers that use ea...   
..                                                                     ...   
145  We have a regular 24-gon with vertices \( S = \{v_0, v_1, \ldots, ...   
146  Let's analyze the problem step-by-step. We have an 11-sided polygo...   
147  We have the sequence defined by \[ x_1 = \frac{25}{11} \] and \[ x...   
148  Given \(\triangle ABC\) with \(\angle A = 90^\circ\), and \(BC = 3...   
149  Given the function \[ f(x) = \frac{(x-18)(x-72)(x-98)(x-k)}{x} \] ...   

             pred_answer  metric  
0                     70  ✔️ [1]  
1                    588  ✔️ [1]  
2                    764  ✔️ [0]  
3                    117  ✔️ [1]  
4                    279  ✔️ [1]  
..                   ...     ...  
145                 4208  ✔️ [0]  
146                   19  ✔️ [1]  
147                  336  ✔️ [0]  
148  \[\n\boxed{104}\n\]  ✔️ [0]  
149                  188  ✔️ [0]  

[150 rows x 5 columns]

EvaluationResult(score=44.67, results=<list of 150 results>)

workflow so far
- create testset -> array of dspy.examples (configure input to "problem" for each example)
- setup program schema (configure input, output with dspy format and set high level prompt)
- configure metric to define output vals to 0 or 1
- setup evaluate()
- run


quick notes
- we define a signature class to add a high level prompt - this provides better context to the llm
- ensure that the variables (question, solution) in signature class are the same as the ones in the dataset
- a simpler way to write signature is dspy.chainofthough("problem -> solution") but it doesnt provide much context


# optimize with GEPA
# GEPA is reflective prompt optimizer, it reflects on its thinking wrt to solutions to frame better prompt

In [14]:
def metric_w_feedback(example, prediction, trace=None, pred_name=None, pred_trace=None):
  correct_answer = int(example["answer"])
  correct_solution = example.get("solution", "")
  try:
    llm_answer = int(prediction.answer)
  except ValueError as e:
    feedback = f"""The final answer must be a valid integer and nothing else. You responded with '{prediction.answer}',
    which couldn't be parsed as a python integer.
    Please ensure your answer is a valid integer without any additional text or formatting. The correct answer is '{correct_answer}'"""
    if correct_solution:
      feedback += f""" Here's the full step-by-step solution:\n{correct_solution}\n\n
      Think about what takeaways you can learn from this solution to improve your future answers and
      approach to similar problems and ensure your final answer is a valid integer."""
    return dspy.Prediction(score = 0, feedback=feedback)

  score = int(correct_answer == llm_answer)

  feedback = ""

  if score == 1:
    feedback = f"Your answer is correct. The correct answer is '{correct_answer}'."
  else:
    feedback = f"Your answer is incorrect. The correct answer is '{correct_answer}'."

  if correct_solution:
    feedback += f""" Here's the full step-by-step solution:\n{correct_solution}\n\n
    Think about what takeaways you can learn from this solution to improve your future answers
    and approach to similar problems."""


  return dspy.Prediction(score=score, feedback=feedback)


quick notes

- we create a new metric that feeds feedback into the llm thinking
- first, we check for any errors in output. if yes, we feeback a string referring correct answer to predicted along with solution (if any)
- next, we check if score is 1 (correct answer) and let it know that it was correct along with solution (if any)
- if score is 0 (wrong answer), we provide the correct answer with reasoning from solutino
- then we return dspy.prediction with updated score and feedback text

In [20]:
from dspy import GEPA

optimizer = GEPA(
    metric=metric_w_feedback,
    num_threads=32,
    auto="light",
    track_stats=True,
    reflection_minibatch_size=3,
    reflection_lm=dspy.LM(model="gpt-5", temperature=1.0, max_tokens=32000, api_key=OPENAI_API_KEY)
)


optimized_program = optimizer.compile(
    program,
    trainset=train_set,
    valset=val_set
  )

2026/01/16 22:51:22 INFO dspy.teleprompt.gepa.gepa: Running GEPA for approx 560 metric calls of the program. This amounts to 6.22 full evals on the train+val set.
2026/01/16 22:51:22 INFO dspy.teleprompt.gepa.gepa: Using 45 examples for tracking Pareto scores. You can consider using a smaller sample of the valset to allow GEPA to explore more diverse solutions within the same budget. GEPA requires you to provide the smallest valset that is just large enough to match your downstream task distribution, while providing as large trainset as possible.
GEPA Optimization:   0%|          | 0/560 [00:00<?, ?rollouts/s]/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content='[[ ## re...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [01:07<00:00, 22.64s/it]

2026/01/16 22:54:30 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content='```\nYou...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(
2026/01/16 22:55:25 INFO dspy.teleprompt.gepa.gepa: Iteration 1: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics problem. Your job is to solve it correctly and output the final result in the required concise format.

Follow these guidelines:

1) Output format
- Provide two sections in order:
  - reaso

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:53<00:00, 17.70s/it]

2026/01/16 23:01:07 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/16 23:02:30 INFO dspy.teleprompt.gepa.gepa: Iteration 2: Proposed new text for predict: You are given a single input field named “problem” that states a contest-style math problem. Your job is to solve it and produce two sections in your output:

- A brief “### reasoning” section with a clear, correct, and efficient solution (showing the key steps, not verbose derivations).
- A concise “### answer” section containing only the final answer in the simplest required form (e.g., an integer or a simplified exact expression), with no extra text.

General requirements:
- Use exact arithmetic and symbolic manipulation whenever possible; avoid decimal approximations unless the problem explicitly asks for them.
- If the problem asks for a remainder modulo m, give the unique nonnegative integer less than m.
- Keep the reasoning succinct but include the critical steps and justifications. Prefer elegant, generalizable arguments when available.
- Do a quick independent check (algebraic, geom

Average Metric: 2.00 / 2 (100.0%):  67%|██████▋   | 2/3 [00:22<00:10, 10.63s/it]

2026/01/16 23:06:39 WARNING dspy.adapters.json_adapter: Failed to use structured output format, falling back to JSON mode.


Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [02:07<00:00, 42.54s/it]

2026/01/16 23:08:21 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content='{\n  "re...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(
2026/01/16 23:09:38 INFO dspy.teleprompt.gepa.gepa: Iteration 3: Proposed new text for predict: You will be given a standalone math problem (often AIME/AMC-style). Your task is to solve it correctly and output the final result in the required numeric format. Follow these guidelines:

General solution and output format
- Read the question carefully and identify exactly what numeric qua

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:33<00:00, 11.27s/it]

2026/01/16 23:14:42 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/16 23:15:22 INFO dspy.teleprompt.gepa.gepa: Iteration 4: Proposed new text for predict: You will be given a standalone math problem (often AIME/AMC-style). Your task is to solve it correctly and output the final result in the required numeric format.

General solution and output format
- Read the question carefully and identify exactly what numeric quantity is requested (e.g., an integer, m+n for a reduced fraction m/n, a probability in lowest terms, or a squared distance).
- Show a concise, correct chain of reasoning. Keep it short but complete (avoid long digressions).
- End your response with the final answer as a single number on its own line, with no extra words or symbols. If the problem asks for m+n where m/n is reduced, reduce the fraction before summing.

Algebra/optimization with rectangular boxes (surface area and volume)
- Let the side lengths be x, y, z > 0. Typical constraints:
  - Surface area S_A: 2(xy + yz + zx) = S_A ⇒ σ2 := xy + yz + zx is fixed.
  - Volume V

Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [01:08<00:00, 22.68s/it]

2026/01/16 23:21:06 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)


2026/01/16 23:22:29 INFO dspy.teleprompt.gepa.gepa: Iteration 5: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics problem. Your job is to solve it correctly and output the final result in the required concise format.

Follow these guidelines:

1) Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as the problem requests.

2) General solving strategy
- Parse the problem carefully and restate key constraints internally.
- Prefer exact algebraic manipulation over numeric approximation. Double-check arithmetic.
- Prove uniqueness or extremality (greatest/smallest) rather than stopping at a candidate. If multiple solutions are possible, count them correctly.
- If the problem asks for “ordered pairs/triples,” count ordered

Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:59<00:00, 19.76s/it]

2026/01/16 23:28:15 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2026/01/16 23:28:15 INFO dspy.teleprompt.gepa.gepa: Iteration 6: All subsample scores perfect. Skipping.
2026/01/16 23:28:15 INFO dspy.teleprompt.gepa.gepa: Iteration 6: Reflective mutation did not propose a new candidate
GEPA Optimization:  54%|█████▍    | 303/560 [36:53<34:43,  8.11s/rollouts]2026/01/16 23:28:15 INFO dspy.teleprompt.gepa.gepa: Iteration 7: Selected program 5 score: 0.5111111111111111



Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [03:38<00:00, 72.89s/it]

2026/01/16 23:31:54 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)


2026/01/16 23:33:18 INFO dspy.teleprompt.gepa.gepa: Iteration 7: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics/geometry problem. Your job is to solve it correctly and output the final result in the required concise format.

Follow these guidelines:

1) Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as the problem requests.
- When the problem asks for m+n (or similar) after expressing a quantity in a specified form (e.g., m/√n, m√n with n squarefree), first put the expression into exactly that form with integer parameters meeting the stated conditions, then compute and output the requested sum. Do not rationalize or reformat into a different form than requested.

2) General solving strategy
- Parse the problem 

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [01:12<00:00, 24.05s/it]

2026/01/16 23:39:02 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/16 23:39:53 INFO dspy.teleprompt.gepa.gepa: Iteration 8: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics/geometry problem. Your job is to solve it correctly and output the final result in the required concise format.

Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as the problem requests.

General solving strategy
- Parse the problem carefully and restate key constraints internally.
- Prefer exact algebraic manipulation over numeric approximation. Double-check arithmetic.
- Prove uniqueness or extremality (greatest/smallest) rather than stopping at a candidate. If multiple solutions are possible, count them correctly.

Useful domain tactics and checks
- Base representation and digit constraints:
  - In b

Average Metric: 0.00 / 3 (0.0%): 100%|██████████| 3/3 [03:08<00:00, 62.78s/it]

2026/01/16 23:43:30 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 3 (0.0%)


/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content="```\nYou...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(
2026/01/16 23:44:46 INFO dspy.teleprompt.gepa.gepa: Iteration 9: Proposed new text for predict: You will be given a standalone math problem (often AIME/AMC-style). Your task is to solve it correctly and output the final result in the required numeric format.

General solution and output format
- Read the question carefully and identify exactly what numeric quantity is requested (e.g.,

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [01:43<00:00, 34.39s/it]

2026/01/16 23:50:30 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/16 23:51:34 INFO dspy.teleprompt.gepa.gepa: Iteration 10: Proposed new text for predict: You are given a single field named "problem" containing a standalone competition-style math problem (algebra/number theory/combinatorics/geometry). Your task is to solve it exactly and return the result in the required concise format.

Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as requested.
- Keep the reasoning focused: show the key steps/identities and the verification; avoid meandering coordinates or unnecessary approximations.

General strategy
1) Parse the problem carefully and identify all constraints (ordered/unordered, integer/nonnegative, base/digit bounds, geometric relations).
2) Prefer exact algebraic manipulation over numeric approximation. Double-check arithmetic and simplify.
3) If

Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:27<00:00,  9.00s/it]

2026/01/16 23:53:30 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2026/01/16 23:53:30 INFO dspy.teleprompt.gepa.gepa: Iteration 11: All subsample scores perfect. Skipping.
2026/01/16 23:53:30 INFO dspy.teleprompt.gepa.gepa: Iteration 11: Reflective mutation did not propose a new candidate
GEPA Optimization:  75%|███████▌  | 420/560 [1:02:08<28:34, 12.24s/rollouts]2026/01/16 23:53:30 INFO dspy.teleprompt.gepa.gepa: Iteration 12: Selected program 1 score: 0.5555555555555556



Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [03:10<00:00, 63.38s/it]

2026/01/16 23:56:40 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/16 23:57:54 INFO dspy.teleprompt.gepa.gepa: Iteration 12: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics/geometry problem. Your job is to solve it correctly and output the final result in the required concise format.

Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or a single expression exactly as requested.

General solving strategy
- Parse the problem carefully and restate key constraints internally.
- Prefer exact algebraic manipulation over numeric approximation. Keep fractions exact; avoid rounding.
- If extremal or uniqueness is involved, justify it structurally or by a complete (but pruned) case analysis.
- If multiple solutions are possible, count them correctly and clarify whether ordered or unordered is intended; a

Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [01:24<00:00, 28.05s/it]

2026/01/16 23:59:46 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)


2026/01/17 00:01:19 INFO dspy.teleprompt.gepa.gepa: Iteration 13: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics/geometry problem. Your job is to solve it correctly and output the final result in the required concise format.

Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as the problem requests.

General solving strategy
- Parse the problem carefully and restate key constraints internally.
- Prefer exact algebraic manipulation over numeric approximation. Double‑check arithmetic.
- Prove uniqueness or extremality (greatest/smallest) rather than stopping at a candidate. If multiple solutions are possible, count them correctly.
- Verify the solution by substituting back into the original constraints.
- For answer

Average Metric: 3.00 / 3 (100.0%): 100%|██████████| 3/3 [00:27<00:00,  9.33s/it]

2026/01/17 00:03:53 INFO dspy.evaluate.evaluate: Average Metric: 3.0 / 3 (100.0%)
2026/01/17 00:03:53 INFO dspy.teleprompt.gepa.gepa: Iteration 14: All subsample scores perfect. Skipping.
2026/01/17 00:03:53 INFO dspy.teleprompt.gepa.gepa: Iteration 14: Reflective mutation did not propose a new candidate
GEPA Optimization:  78%|███████▊  | 435/560 [1:12:30<42:12, 20.26s/rollouts]2026/01/17 00:03:53 INFO dspy.teleprompt.gepa.gepa: Iteration 15: Selected program 5 score: 0.5111111111111111



Average Metric: 1.00 / 3 (33.3%): 100%|██████████| 3/3 [01:13<00:00, 24.40s/it]

2026/01/17 00:05:06 INFO dspy.evaluate.evaluate: Average Metric: 1.0 / 3 (33.3%)


2026/01/17 00:06:22 INFO dspy.teleprompt.gepa.gepa: Iteration 15: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics/geometry problem. Solve it exactly and output in the required concise format.

1) Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as requested.

2) General solving strategy
- Parse the problem carefully; restate key constraints to yourself.
- Prefer exact algebraic manipulation and exact arithmetic; avoid unnecessary approximations.
- For extremal/uniqueness questions, justify optimality or uniqueness.
- If multiple solutions are possible, count them correctly; mind ordered vs unordered as stated.
- Always substitute back or check conditions (digit/base bounds, monotonicity, intersection constraints, 

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:27<00:00,  9.06s/it]

2026/01/17 00:10:40 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/17 00:12:08 INFO dspy.teleprompt.gepa.gepa: Iteration 16: Proposed new text for predict: You are given a single field named "problem" containing a standalone math/number theory/combinatorics (and occasionally geometry) problem. Your job is to solve it correctly and output the final result in the required concise format.

Follow these guidelines:

1) Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, leading zeros, or extra text)
- The final answer must be a single value or expression exactly as the problem requests.

2) General solving strategy
- Parse the problem carefully and restate key constraints internally.
- Prefer exact algebraic manipulation over numeric approximation. Keep fractions reduced; simplify radicals. Double-check arithmetic and algebraic transformations.
- Prove uniqueness or extremality (greatest/smallest) rather than stopping at a candidate. If mu

Average Metric: 2.00 / 3 (66.7%): 100%|██████████| 3/3 [00:30<00:00, 10.06s/it]

2026/01/17 00:16:17 INFO dspy.evaluate.evaluate: Average Metric: 2.0 / 3 (66.7%)


2026/01/17 00:17:24 INFO dspy.teleprompt.gepa.gepa: Iteration 17: Proposed new text for predict: You are given a single field named "problem" containing one standalone math/number theory/combinatorics/geometry problem. Your job is to solve it correctly and output the final result in the required concise format.

Output format
- Provide two sections in order:
  - reasoning: a concise but complete derivation
  - answer: the final result only (no units, labels, punctuation, or extra text)
- The final answer must be a single value or expression exactly as the problem requests.

General solving strategy
- Parse the problem carefully and restate key constraints internally (ranges, integrality, base/digit bounds, geometric configurations).
- Prefer exact symbolic/algebraic manipulation over numeric approximation. Keep fractions exact and rationalize when needed.
- Prove uniqueness or extremality (greatest/smallest) rather than stopping at a candidate. If multiple solutions are possible, count

In [21]:
program.predict.signature.instructions

'Solve the problem and provide the answer in the correct format.'

In [22]:
optimized_program.predict.signature.instructions

'You are given a single field named "problem" containing a standalone math/number theory/combinatorics problem. Your job is to solve it correctly and output the final result in the required concise format.\n\nFollow these guidelines:\n\n1) Output format\n- Provide two sections in order:\n  - reasoning: a concise but complete derivation\n  - answer: the final result only (no units, labels, punctuation, or extra text)\n- The final answer must be a single value or expression exactly as the problem requests.\n\n2) General solving strategy\n- Parse the problem carefully and restate key constraints internally.\n- Prefer exact algebraic manipulation over numeric approximation. Double-check arithmetic.\n- Prove uniqueness or extremality (greatest/smallest) rather than stopping at a candidate. If multiple solutions are possible, count them correctly.\n\n3) Useful domain tactics and checks\n- Base representation and digit constraints:\n  - In base-k, every digit must be in {0, 1, ..., k-1}. Lead

In [23]:
evaluate(optimized_program)

Average Metric: 1.00 / 1 (100.0%):   1%|          | 1/150 [00:05<12:38,  5.09s/it]

/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content='[[ ## re...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(


Average Metric: 5.00 / 5 (100.0%):   3%|▎         | 4/150 [00:10<04:48,  1.98s/it]

/usr/local/lib/python3.12/dist-packages/pydantic/main.py:464: UserWarning: Pydantic serializer warnings:
  PydanticSerializationUnexpectedValue(Expected 10 fields but got 6: Expected `Message` - serialized value may not be as expected [field_name='message', input_value=Message(content="[[ ## re...: None}, annotations=[]), input_type=Message])
  PydanticSerializationUnexpectedValue(Expected `StreamingChoices` - serialized value may not be as expected [field_name='choices', input_value=Choices(finish_reason='st...ider_specific_fields={}), input_type=Choices])
  return self.__pydantic_serializer__.to_python(


Average Metric: 69.00 / 150 (46.0%): 100%|██████████| 150/150 [03:19<00:00,  1.33s/it]

2026/01/17 00:34:25 INFO dspy.evaluate.evaluate: Average Metric: 69 / 150 (46.0%)


problem  \
0    Find the sum of all integer bases $b>9$ for which $17_b$ is a divi...   
1    On $\triangle ABC$ points $A, D, E$, and $B$ lie in that order on ...   
2    The 9 members of a baseball team went to an ice-cream parlor after...   
3    Find the number of ordered pairs $(x,y)$, where both $x$ and $y$ a...   
4    There are $8!= 40320$ eight-digit positive integers that use each ...   
..                                                                     ...   
145  Let $S$ be the set of vertices of a regular $24$-gon. Find the num...   
146  Let $A_1 A_2 A_3 \ldots A_{11}$ be an $11$-sided non-convex simple...   
147  Let $x_1, x_2, x_3, \ldots$ be a sequence of rational numbers defi...   
148  Let $\triangle ABC$ be a right triangle with $\angle A = 90^\circ$...   
149  There are exactly three positive real numbers $k$ such that the fu...   

     example_answer  \
0                70   
1               588   
2                16   
3               117   
4               279   
..              ...   
145             113   
146              19   
147             248   
148             104   
149             240   

                                                                 reasoning  \
0    First, interpret the given numbers in base \( b \): - \(17_b = 1 \...   
1    First, we set coordinate axes and place points to utilize given di...   
2    We have 9 players, each choosing a flavor from {Chocolate (C), Van...   
3    We are given the equation \[ 12x^2 - xy - 6y^2 = 0, \] with intege...   
4    We are looking at the set of all 8-digit numbers that use each dig...   
..                                                                     ...   
145  We are given a regular 24-gon with vertices \( S \), and we want t...   
146  We have an 11-sided polygon \( A_1 A_2 \cdots A_{11} \) with the f...   
147  Given the sequence: \[ x_1 = \frac{25}{11}, \quad x_{k+1} = \frac{...   
148  Given right triangle \( \triangle ABC \) with \(\angle A = 90^\cir...   
149  Consider the function \[ f(x) = \frac{(x - 18)(x - 72)(x - 98)(x -...   

    pred_answer  metric  
0            70  ✔️ [1]  
1           588  ✔️ [1]  
2            16  ✔️ [1]  
3           117  ✔️ [1]  
4          -297  ✔️ [0]  
..          ...     ...  
145        4464  ✔️ [0]  
146          19  ✔️ [1]  
147         108  ✔️ [0]  
148         104  ✔️ [1]  
149         188  ✔️ [0]  

[150 rows x 5 columns]

EvaluationResult(score=46.0, results=<list of 150 results>)

In [24]:
evaluate(optimized_program)

Average Metric: 65.00 / 150 (43.3%): 100%|██████████| 150/150 [00:00<00:00, 173.31it/s]


2026/01/17 00:37:31 INFO dspy.evaluate.evaluate: Average Metric: 65 / 150 (43.3%)


problem  \
0    Find the sum of all integer bases $b>9$ for which $17_b$ is a divi...   
1    On $\triangle ABC$ points $A, D, E$, and $B$ lie in that order on ...   
2    The 9 members of a baseball team went to an ice-cream parlor after...   
3    Find the number of ordered pairs $(x,y)$, where both $x$ and $y$ a...   
4    There are $8!= 40320$ eight-digit positive integers that use each ...   
..                                                                     ...   
145  Let $S$ be the set of vertices of a regular $24$-gon. Find the num...   
146  Let $A_1 A_2 A_3 \ldots A_{11}$ be an $11$-sided non-convex simple...   
147  Let $x_1, x_2, x_3, \ldots$ be a sequence of rational numbers defi...   
148  Let $\triangle ABC$ be a right triangle with $\angle A = 90^\circ$...   
149  There are exactly three positive real numbers $k$ such that the fu...   

     example_answer  \
0                70   
1               588   
2                16   
3               117   
4               279   
..              ...   
145             113   
146              19   
147             248   
148             104   
149             240   

                                                                 reasoning  \
0    First, translate the given numbers from base \( b \) to decimal ex...   
1    First, we set coordinate axes and place points to utilize given di...   
2    We have 9 players choosing among 3 flavors: chocolate (C), vanilla...   
3    We want the integer ordered pairs \((x,y)\) with \(-100 \leq x,y \...   
4    We consider the 8-digit integers formed by the digits 1 through 8 ...   
..                                                                     ...   
145  We are given a regular 24-gon with vertices \( S \), and we want t...   
146  We have an 11-gon \( A_1 A_2 \ldots A_{11} \) with the following p...   
147  Given the recurrence: \[ x_{k+1} = \frac{1}{3}\left(x_k + \frac{1}...   
148  Given: \(\triangle ABC\) is right-angled at \(A\) with \(BC = 38\)...   
149  Start with the given function: \[ f(x) = \frac{(x - 18)(x - 72)(x ...   

    pred_answer  metric  
0            70  ✔️ [1]  
1           588  ✔️ [1]  
2            16  ✔️ [1]  
3           119  ✔️ [0]  
4           279  ✔️ [1]  
..          ...     ...  
145        4464  ✔️ [0]  
146          64  ✔️ [0]  
147         722  ✔️ [0]  
148          98  ✔️ [0]  
149         188  ✔️ [0]  

[150 rows x 5 columns]

EvaluationResult(score=43.33, results=<list of 150 results>)

In [25]:
evaluate(optimized_program)

Average Metric: 65.00 / 150 (43.3%): 100%|██████████| 150/150 [00:00<00:00, 157.71it/s]

2026/01/17 00:40:08 INFO dspy.evaluate.evaluate: Average Metric: 65 / 150 (43.3%)


problem  \
0    Find the sum of all integer bases $b>9$ for which $17_b$ is a divi...   
1    On $\triangle ABC$ points $A, D, E$, and $B$ lie in that order on ...   
2    The 9 members of a baseball team went to an ice-cream parlor after...   
3    Find the number of ordered pairs $(x,y)$, where both $x$ and $y$ a...   
4    There are $8!= 40320$ eight-digit positive integers that use each ...   
..                                                                     ...   
145  Let $S$ be the set of vertices of a regular $24$-gon. Find the num...   
146  Let $A_1 A_2 A_3 \ldots A_{11}$ be an $11$-sided non-convex simple...   
147  Let $x_1, x_2, x_3, \ldots$ be a sequence of rational numbers defi...   
148  Let $\triangle ABC$ be a right triangle with $\angle A = 90^\circ$...   
149  There are exactly three positive real numbers $k$ such that the fu...   

     example_answer  \
0                70   
1               588   
2                16   
3               117   
4               279   
..              ...   
145             113   
146              19   
147             248   
148             104   
149             240   

                                                                 reasoning  \
0    First, translate the given numbers from base \( b \) to decimal ex...   
1    First, we set coordinate axes and place points to utilize given di...   
2    We have 9 players choosing among 3 flavors: chocolate (C), vanilla...   
3    We want the integer ordered pairs \((x,y)\) with \(-100 \leq x,y \...   
4    We consider the 8-digit integers formed by the digits 1 through 8 ...   
..                                                                     ...   
145  We are given a regular 24-gon with vertices \( S \), and we want t...   
146  We have an 11-gon \( A_1 A_2 \ldots A_{11} \) with the following p...   
147  Given the recurrence: \[ x_{k+1} = \frac{1}{3}\left(x_k + \frac{1}...   
148  Given: \(\triangle ABC\) is right-angled at \(A\) with \(BC = 38\)...   
149  Start with the given function: \[ f(x) = \frac{(x - 18)(x - 72)(x ...   

    pred_answer  metric  
0            70  ✔️ [1]  
1           588  ✔️ [1]  
2            16  ✔️ [1]  
3           119  ✔️ [0]  
4           279  ✔️ [1]  
..          ...     ...  
145        4464  ✔️ [0]  
146          64  ✔️ [0]  
147         722  ✔️ [0]  
148          98  ✔️ [0]  
149         188  ✔️ [0]  

[150 rows x 5 columns]

EvaluationResult(score=43.33, results=<list of 150 results>)

In [26]:
with dspy.context(cache=False):
    evaluate(optimized_program)

Average Metric: 65.00 / 150 (43.3%): 100%|██████████| 150/150 [00:04<00:00, 30.79it/s]

2026/01/17 00:48:46 INFO dspy.evaluate.evaluate: Average Metric: 65 / 150 (43.3%)


problem  \
0    Find the sum of all integer bases $b>9$ for which $17_b$ is a divi...   
1    On $\triangle ABC$ points $A, D, E$, and $B$ lie in that order on ...   
2    The 9 members of a baseball team went to an ice-cream parlor after...   
3    Find the number of ordered pairs $(x,y)$, where both $x$ and $y$ a...   
4    There are $8!= 40320$ eight-digit positive integers that use each ...   
..                                                                     ...   
145  Let $S$ be the set of vertices of a regular $24$-gon. Find the num...   
146  Let $A_1 A_2 A_3 \ldots A_{11}$ be an $11$-sided non-convex simple...   
147  Let $x_1, x_2, x_3, \ldots$ be a sequence of rational numbers defi...   
148  Let $\triangle ABC$ be a right triangle with $\angle A = 90^\circ$...   
149  There are exactly three positive real numbers $k$ such that the fu...   

     example_answer  \
0                70   
1               588   
2                16   
3               117   
4               279   
..              ...   
145             113   
146              19   
147             248   
148             104   
149             240   

                                                                 reasoning  \
0    First, translate the given numbers from base \( b \) to decimal ex...   
1    First, we set coordinate axes and place points to utilize given di...   
2    We have 9 players choosing among 3 flavors: chocolate (C), vanilla...   
3    We want the integer ordered pairs \((x,y)\) with \(-100 \leq x,y \...   
4    We consider the 8-digit integers formed by the digits 1 through 8 ...   
..                                                                     ...   
145  We are given a regular 24-gon with vertices \( S \), and we want t...   
146  We have an 11-gon \( A_1 A_2 \ldots A_{11} \) with the following p...   
147  Given the recurrence: \[ x_{k+1} = \frac{1}{3}\left(x_k + \frac{1}...   
148  Given: \(\triangle ABC\) is right-angled at \(A\) with \(BC = 38\)...   
149  Start with the given function: \[ f(x) = \frac{(x - 18)(x - 72)(x ...   

    pred_answer  metric  
0            70  ✔️ [1]  
1           588  ✔️ [1]  
2            16  ✔️ [1]  
3           119  ✔️ [0]  
4           279  ✔️ [1]  
..          ...     ...  
145        4464  ✔️ [0]  
146          64  ✔️ [0]  
147         722  ✔️ [0]  
148          98  ✔️ [0]  
149         188  ✔️ [0]  

[150 rows x 5 columns]

Final thoughts

there was a slight improvement in performance from 44.67 to 46

this is not great but could also be because
- the dataset is really small and less diverse
- the llm is not the best and prompting can only do so much


Cons
- the optimisation process took > 60 min for a ~1% increase in performance


finally, I for this specific experiment, I didnt think it was worth the tokens for a 2% imrprovement but according to gepa results in their docs, it seems really useful

next time
- Ill consider better lengthy dataset so it has diverse problems during training
- or consider a smaller dataset that contains all the diverse topics in the dataset